### PR TITLE
LOG 1838: Adding Promtail, Fluentd, and Openshift Logging docs

### PR DESCRIPTION
**Pull request by Red-GV**
This PR adds documentation for using the Promtail and Openshift-Logging forwarder for sending logs into the gateway.

JIRA: https://issues.redhat.com/browse/LOG-1838

### CODE REVIEW

File: docs/forwarding_logs_to_gateway.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Forwarding Logs to Loki Operator
+
+This document will describe how to send application, infrastructure, and audit logs to the Lokistack Gateway as different tenants using Promtail. The gateway provides secure access to the distributor (and query-frontend) via consulting an OAuth/OIDC endpoint for the request subject.
+
+__Please read the `hacking_loki_operator.md` document before proceeding with the following instructions.__
```
Comment (sasagarw) [inline docs/forwarding_logs_to_gateway.md:5]: Linking the document link would be good here.
```suggestion
__Please read the [hacking_loki_operator.md](https://github.com/ViaQ/loki-operator/blob/master/docs/hack_loki_operator.md) document before proceeding with the following instructions.__
```

File: docs/forwarding_logs_to_gateway.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+Support will be added in the near future.
+
+## Promtail
+
+[Promtail](https://grafana.com/docs/loki/latest/clients/promtail/) is an agent managed by Grafana which forwards logs to a Loki instance.
+
+In order to generate an instance of Promtail with the necessary authorization (service account token) and authentication (rbac) resources for interacting with the gateway, perform the following steps:
+
+1. Deploy the Loki Operator to the cluster with the gateway component
+2. Execute the following commands in the terminal:
```
Comment (sasagarw) [inline docs/forwarding_logs_to_gateway.md:22]:
```suggestion
2. Execute the following command in the terminal:
```

**Review by sasagarw — CHANGES_REQUESTED** [review]
Looks good in general. However, there are some missing bits which I can see:
1. To add what should we expect to see after we have applied this `hack/promtail_client.yaml`. A short paragraph or some kubectl commands which will show us the deployment so that we know what is expected to happen after this.
2. Link the docs into index.md to show up here: https://viaq.github.io/loki-operator/

**Review by periklis — CHANGES_REQUESTED** [review]
This PR is not sufficiently matching the task's acceptance criteria:
1. Regardless OpenShift Logging, we should add a section for configuring fluentd with the grafana fluentd loki plugin to visualize how to achieve a similar behavior.
2. The manifests should not be put into the hack folder. This nurtures the expectation that these manifests are actively maintained in addition to the lokistack/loki-operator. Please add any manifests inline in splitting them per manifest and adding some description on why we need this and where to put attention on customization.
3. For better visualization the entrypoint site should like to two different guides one for fluentd and one for promtail, because the manifests are quite noisy.

File: docs/forwarding_logs_to_gateway.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# Forwarding Logs to Loki Operator
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:1]: Suggest `Forwardig logs to LokiStack`

File: docs/forwarding_logs_to_gateway.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+This document will describe how to send application, infrastructure, and audit logs to the Lokistack Gateway as different tenants using Promtail. The gateway provides secure access to the distributor (and query-frontend) via consulting an OAuth/OIDC endpoint for the request subject.
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:3]: `Using Promtail or Fluentd`

Comment (periklis) [reply docs/forwarding_logs_to_gateway.md:3]: `The built-in gateway provides..`

File: docs/forwarding_logs_to_gateway.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+__Please read the [hacking_loki_operator.md](./hacking_loki_operator.md) document before proceeding with the following instructions.__
+
+_Note: While this document will only give instructions for two methods of log forwarding into the gateway, the examples given in the Promtail section can be extrapolated to other log forwarders._
+
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:8]: `in the Promtail and Fluentd sections`

File: docs/forwarding_logs_to_gateway.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+Although there is a way to [forward logs to an external Loki instance](https://docs.openshift.com/container-platform/4.9/logging/cluster-logging-external.html#cluster-logging-collector-log-forward-loki_cluster-logging-external), [Openshift Logging](https://github.com/openshift/cluster-logging-operator) does not currently have support to send logs through the Lokistack Gateway.
+
+Support will be added in the near future.
+
+## Forwarding Clients
+
+In order to enable communication between the client(s) and the gateway, follow these steps:
+
+1. Deploy the Loki Operator and an `lokistack` instance with the gateway enabled.
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:19]: Need to precise what `with the gateway enabled` means, i.e. your operator should run with the appropriate feature flags.

File: docs/forwarding_logs_to_gateway.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+kind: ClusterRoleBinding
+metadata:
+  name: lokistack-dev-tenant-logs-binding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: lokistack-dev-tenant-logs-role
+subjects:
+- kind: User
+  name: "kube:admin"
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:60]: I suggest to drop the `kube:admin` user here. It is a temporary user account that is usually dropped after configuring IDP on k8s.

File: docs/forwarding_logs_to_gateway.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  @type loki
+  # ...
+  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
+  url: http://lokistack-gateway-http-lokistack-dev.openshift-logging.svc:8080/api/logs/v1/application
+</match>
+```
+
+## Troubleshooting
+
+### Log Entries Out of Order
```
Comment (periklis) [inline docs/forwarding_logs_to_gateway.md:104]: This depends highly on the t-shirt size you pick. IIRC `1x.medium` can handle `IngestionRate: 10` and `IngestionBurstSize: 20`. Right?

**Review by sasagarw — CHANGES_REQUESTED** [review]
Looks good. However, I still miss the step-by-step guide to use OpenShift Logging ClusterLogForwarder part which is mentioned in AC of this task. Have we updated the AC?

File: index.md [hunk 9]
```diff
@@ -8,6 +8,10 @@ This is the Kubernetes Operator for Loki provided by the Red Hat OpenShift engin
 * Also, there is a step-by-step guide on how to test Loki-operator on [OpenShift](https://github.com/ViaQ/loki-operator/blob/master/docs/hack_loki_operator.md#hacking-on-loki-operator-on-openshift).
 * There is also a [basic troubleshooting guide](https://github.com/ViaQ/loki-operator/blob/master/docs/hack_loki_operator.md#basic-troubleshooting-on-hacking-on-loki-operator) if you run into some common problems.
 
+### Sending Logs to Loki through the Gateway Component
+
+* The [forwarding_logs_to_gateway.md](https://github.com/ViaQ/loki-operator/tree/master/docs/forwarding_logs_to_gateway.md) document provides instructions for configuring forwarding clients to ship logs to Loki through the gateway component.
```
[the remark this answers is not in the document]
Comment (periklis) [reply index.md:13]: Yes that would make it easier to identify the client.

**Comment by Red-GV** [comment]
@sasagarw The problem is that we currently do not have a path in with the ClusterLogForwarder because it currently does not support using the bearer token for authorization. I could toss in what it would look like in theory, but currently we can't actually use it yet.

**Review by periklis — APPROVED** [review]
LGTM

**Pull request merged**